velero-csi-plugin/csi_snapshotter.go

```diff
@@ -45,55 +54,75 @@ type CSISnapshotter struct {
 func (p *CSISnapshotter) AppliesTo() (velero.ResourceSelector, error) {
 	p.log.Info("CSISnapshotter AppliesTo")
 
+	//TODO: Update this to look at only PVCs with "magic" label
 	return velero.ResourceSelector{
 		IncludedResources: []string{"persistentvolumeclaims"},
 	}, nil
 }
 
 // Execute allows the ItemAction to perform arbitrary logic with the item being backed up,
 // in this case, setting a custom annotation on the item being backed up.
-func (p *CSISnapshotter) Execute(item runtime.Unstructured, backup *v1.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+func (p *CSISnapshotter) Execute(item runtime.Unstructured, backup *velerov1api.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
 	p.log.Info("CSISnapshotter Execute")
 
+	// Do nothing if volume snapshots have not been requested in this backup
+	if boolptr.IsSetToFalse(backup.Spec.SnapshotVolumes) {
+		p.log.Infof("Volume snapshotting not requested for backup %s/%s", backup.Namespace, backup.Name)
+		return item, nil, nil
+	}
+
 	var pvc corev1api.PersistentVolumeClaim
 	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &pvc); err != nil {
 		return nil, nil, errors.WithStack(err)
 	}
 
-	// no storage class: we don't know how to map to a VolumeSnapshotClass
-	if pvc.Spec.StorageClassName == nil {
-		return item, nil, nil
-	}
-
 	client, snapshotClient, err := getClients()
 	if err != nil {
 		return nil, nil, errors.WithStack(err)
 	}
 
-	storageClass, err := client.StorageV1().StorageClasses().Get(*pvc.Spec.StorageClassName, metav1.GetOptions{})
+	// Do nothing if restic is used to backup this PV
+	isResticUsed, err := isPVCBackedUpByRestic(pvc.Namespace, pvc.Name, client.CoreV1())
 	if err != nil {
-		return nil, nil, errors.Wrap(err, "error getting storage class")
+		return nil, nil, errors.WithStack(err)
 	}
-
-	snapshotClasses, err := snapshotClient.SnapshotV1beta1().VolumeSnapshotClasses().List(metav1.ListOptions{})
+	if isResticUsed {
+		p.log.Infof("Nothing to do for PVC %s/%s is backed up using restic")
```

Comment:
Nit: let's say something like "Skipping CSI snapshot of PVC %s/%s because it's being backed up with restic" to be a little more clear.

Although, now that I'm reading L95-102, I guess we don't even know that this *is* a CSI PVC yet, so my suggestion may also be misleading.

Perhaps we should do the CSI check first, so we can exit right away if it's not a CSI volume. WDYT?

Reply:
I agree w/ changing the order of the checks. The `isCSIPV` check is less expensive of the two calls.